### PR TITLE
Bluetooth: ISO: Add check for maximum ISO SDU size

### PR DESCRIPTION
**Pull request by Thalley**
Verify that the HCI ISO data packet is smaller than
the maximum SDU size we support, based on the group
settings.

Signed-off-by: Emil Gydesen <emil.gydesen@nordicsemi.no>

### CODE REVIEW

**Comment by Thalley** [comment]
@cvinayak @Casper-Bonde-Bose Do you know if there's any way we can reconstruct the maximum SDU for the ISO peripheral? 

The CIS established event only provides the PDU size, even though it is sent as part of the LL_CIS_REQ: 
image

**Review by Frodevan — CHANGES_REQUESTED** [review]

File: samples/bluetooth/unicast_audio_server/src/main.c [hunk 1]
```diff
@@ -202,16 +203,15 @@ static void audio_timer_timeout(struct k_work *work)
 			net_buf_unref(buf);
 		} else {
 			printk("Sending mock data with len %zu on streams[%zu] (%p)\n",
-			       len_to_send, i, stream);
+			       source_streams[i].len_to_send, i, stream);
+		}
+
+		if (source_streams[i].len_to_send > source_streams[i].max_sdu) {
```
[the remark this answers is not in the document]
Comment (Thalley) [reply samples/bluetooth/unicast_audio_server/src/main.c]: Good catch! Fixed

**Review by Frodevan — DISMISSED** [review]

**Review by alwa-nordic — DISMISSED** [review]

File: subsys/bluetooth/host/iso.c [hunk 2]
```diff
@@ -743,6 +775,13 @@ int bt_iso_chan_send(struct bt_iso_chan *chan, struct net_buf *buf,
 		return -EINVAL;
 	}
 
+	max_data_len = iso_chan_max_data_len(chan, ts);
+	if (buf->len > max_data_len) {
+		BT_DBG("Cannot send %u octets, maximum %u",
+		       buf->len, max_data_len);
+		return -EINVAL;
```
Comment (jhedberg) [inline subsys/bluetooth/host/iso.c:782]: Is `EINVAL` really the most suitable error here? Why not e.g. `EMSGSIZE`?

**Review by jhedberg — CHANGES_REQUESTED** [review]

File: subsys/bluetooth/host/iso.c [hunk 3]
```diff
@@ -705,9 +705,41 @@ void bt_iso_recv(struct bt_conn *iso, struct net_buf *buf, uint8_t flags)
 #endif /* CONFIG_BT_ISO_UNICAST) || defined(CONFIG_BT_ISO_SYNC_RECEIVER */
 
 #if defined(CONFIG_BT_ISO_UNICAST) || defined(CONFIG_BT_ISO_BROADCASTER)
+static uint16_t iso_chan_max_data_len(const struct bt_iso_chan *chan,
+				      uint32_t ts)
+{
+	uint16_t max_data_len;
+
+	if (chan->qos->tx == NULL) {
```
Comment (MariuszSkamra) [inline subsys/bluetooth/host/iso.c:713]:
```suggestion
	if (chan->qos->tx == NULL || bt_dev.le.iso_mtu < BT_HCI_ISO_DATA_HDR_SIZE) {
```

File: subsys/bluetooth/host/iso.c [hunk 4]
```diff
@@ -705,9 +705,41 @@ void bt_iso_recv(struct bt_conn *iso, struct net_buf *buf, uint8_t flags)
...
+					   max_data_len);
+		}
+	} else {
+		if (bt_dev.le.iso_mtu < BT_HCI_ISO_TS_DATA_HDR_SIZE) {
+			max_data_len = 0;
+		} else {
+			max_data_len = MIN(bt_dev.le.iso_mtu - BT_HCI_ISO_TS_DATA_HDR_SIZE,
+					   max_data_len);
+		}
+	}
```
Comment (MariuszSkamra) [inline subsys/bluetooth/host/iso.c:734]:
```suggestion
	if (ts == BT_ISO_TIMESTAMP_NONE) {
		max_data_len = MIN(bt_dev.le.iso_mtu - BT_HCI_ISO_DATA_HDR_SIZE,
				   max_data_len);
	} else {
		max_data_len = MIN(bt_dev.le.iso_mtu - BT_HCI_ISO_TS_DATA_HDR_SIZE,
				   max_data_len);
	}
```

File: subsys/bluetooth/audio/unicast_client.c [hunk 5]
```diff
@@ -575,7 +581,7 @@ static void unicast_client_ep_qos_state(struct bt_audio_ep *ep,
 	qos = net_buf_simple_pull_mem(buf, sizeof(*qos));
 
 	/* Reset any existing QoS configuration */
```
Comment (MariuszSkamra) [inline subsys/bluetooth/audio/unicast_client.c:583]:
```suggestion
	/* Update any existing QoS configuration */
```

**Review by MariuszSkamra — COMMENTED** [review]
Minor suggestion to avoid code duplicates

File: subsys/bluetooth/host/iso.c (hunk 4 above)
Comment (Thalley) [reply subsys/bluetooth/host/iso.c:734]: Hmm, I see what you are trying here (i.e. move the `BT_HCI_ISO_DATA_HDR_SIZE` check to earlier), but we will need two checks for that, as `BT_HCI_ISO_DATA_HDR_SIZE` and `BT_HCI_ISO_TS_DATA_HDR_SIZE` aren't the same size. 

The suggestion you provide here, with https://github.com/zephyrproject-rtos/zephyr/pull/48068/files#r935272842, doesn't work as `bt_dev.le.iso_mtu - BT_HCI_ISO_TS_DATA_HDR_SIZE` may underflow.

File: subsys/bluetooth/host/iso.c (hunk 2 above)
Comment (Thalley) [reply subsys/bluetooth/host/iso.c]: Agreed, `EMSGSIZE` is better here.

File: subsys/bluetooth/host/iso.c (hunk 4 above)
Comment (Thalley) [reply subsys/bluetooth/host/iso.c]: Updated after realizing that we actually do support fragmentation, so this part has been simplified.

**Review by jhedberg — DISMISSED** [review]

**Comment by Casper-Bonde-Bose** [comment]
> @cvinayak @Casper-Bonde-Bose Do you know if there's any way we can reconstruct the maximum SDU for the ISO peripheral?
> 
> The CIS established event only provides the PDU size, even though it is sent as part of the LL_CIS_REQ: image

Sorry, I missed this post.
I believe the Max SDU can be derived approximately:
` (BN * (Max_PDU - 1)) < MAX_SDU <= (BN*Max_PDU)`
This is unless the central controller do not care about air-time efficiency (or knows another configuration is more efficient for the use-case) and simply sets `Max_PDU `to 251 when the SDU is larger than 251. I do not know if there is a specification requirement for the controller to set the `Max_PDU `size to `ceil(SDU/BN)`.
In other words I believe using the GATT layer exchanged values is the correct way of doing it - which is also what this change does..

File: subsys/bluetooth/host/iso.c [hunk 6]
```diff
@@ -705,9 +705,32 @@ void bt_iso_recv(struct bt_conn *iso, struct net_buf *buf, uint8_t flags)
...
+
+	max_data_len = chan->qos->tx->sdu;
+
+	if (ts == BT_ISO_TIMESTAMP_NONE) {
+		max_data_len = MIN(bt_dev.le.iso_mtu - BT_HCI_ISO_DATA_HDR_SIZE,
+				   max_data_len);
+	} else {
+		max_data_len = MIN(bt_dev.le.iso_mtu - BT_HCI_ISO_TS_DATA_HDR_SIZE,
+				   max_data_len);
```
Comment (Casper-Bonde-Bose) [inline subsys/bluetooth/host/iso.c:724]: I always mix up MTU, PDU and SDU sizes, hence correct me if I'm wrong. But I think MTU maps to PDU, and you can have more PDU's to form a SDU. Hence I don't think this check is needed - as max_data_len refers to the SDU and not the PDU. 
What we need to check is that the SDU do not exceed the ISO_buffer_size * ISO_buffer_count from the read_buffer_size_V2 HCI command. To my knowledge there is no way for a controller to inform the host about it not being able to handle SDU's above certain size should it be smaller than the buffers it have.
If I got this detail wrong, please ping me and I'll approve.

**Review by Casper-Bonde-Bose — CHANGES_REQUESTED** [review]
There may also be a host stack configuration that prevents SDU's above a certain size - such as the CONFIG_BT_ISO_TX_MTU.
Note that for this it seems MTU maps to SDU size, where as for the ISO buffer size MTU maps to fragment (PDU) size.

File: subsys/bluetooth/host/iso.c (hunk 6 above)
Comment (Casper-Bonde-Bose) [reply subsys/bluetooth/host/iso.c:724]: There is actually another requirement.
IMO the controller ISO buffers(bt_hci_rp_le_read_buffer_size_v2) also needs to be able to hold BN SDU's - hence if BN is 3 and SDU is 300 and the controller iso_max_len is 256 you would need the controller to have 6 ISO TX buffer entries.
`BN * ceil(SDU/iso_max_len)` needs to be <= `iso_max_num`

Comment (Thalley) [reply subsys/bluetooth/host/iso.c:724]: > I always mix up MTU, PDU and SDU sizes, hence correct me if I'm wrong. But I think MTU maps to PDU, and you can have more PDU's to form a SDU. Hence I don't think this check is needed - as max_data_len refers to the SDU and not the PDU.

Using the term MTU here is a bit odd, as it does correspond to anything in the spec. However it is currently used to cover the `ISO_SDU_Length` of the HCI ISO Data Packet: 
image


i.e. the maximum size of the actual data ISO data. That is why, when allocating the buffer, one should use `BT_ISO_SDU_BUF_SIZE` to also handle the ISO Header (Connection_Handle and ISO_Data_Load_Length)  and the ISO SDU header (Time_Stamp, Packet_Sequence_Number, ISO_SDU_Length and the Packet_Status_Flag). 

So we have SDU_Size (BT_ISO_SDU_BUF_SIZE) = MTU size (CONFIG_BT_ISO_TX_MTU) + 12

The PDU size is the maximum size the controller supports sending over air. This value is limited to maximum 251 (whereas SDU size is limited to 4095). 

The `bt_dev.le.iso_mtu` (perhaps misleading name) is then the **buffer** size read from the controller. This may be different from the PDU size, and is used to determine how data is (and if) fragmented when sending ISO data over HCI to the controller. The check your commented on here should actually be removed, and instead be replaced by a check to see if SDU > bt_dev.le.iso_mtu * bt_dev.le.iso_pkts.limit, to determine if is possible to send the data.

My understanding is that we should not worry about BN; this is a controller thing that is unrelated to how the host sends data, or how the buffers are used.

Comment (Casper-Bonde-Bose) [reply subsys/bluetooth/host/iso.c]: It is correct BN is a controller thing, but if BN=3 for your transport and there is not enough buffers to hold data for a full ISO event inside the controller I fear it will not be able to transmit the data (as the host cannot deliver the data in time).  At least some controllers wants to be able to schedule a full ISO event up front - and will not accept new data arriving within an ISO event.

**Review by Casper-Bonde-Bose — APPROVED** [review]
LGTM

File: subsys/bluetooth/host/iso.c (hunk 6 above)
Comment (Thalley) [reply subsys/bluetooth/host/iso.c]: > At least some controllers wants to be able to schedule a full ISO event up front - and will not accept new data arriving within an ISO event.

That's a good point. I would have expected that a controller would not attempt to send any data until the entire SDU has been received from the host (i.e. either a `BT_ISO_SINGLE` or a `BT_ISO_END` has been received), but that may not be the case (haven't checked/read on this).

**Review by carlescufi — APPROVED** [review]
